app/src/main/java/de/westnordost/streetcomplete/data/maptiles/MapTilesDownloader.kt

```diff
@@ -7,90 +7,90 @@ import de.westnordost.streetcomplete.ktx.format
-                        }
-                    }
+        downloadTilesFlow(bbox).collect { result ->
+            ++tileCount
+            when (result) {
+                is DownloadFailure -> ++failureCount
+                is DownloadSuccess -> {
+                    if (result.alreadyCached) cachedSize += result.size
+                    else downloadedSize += result.size
                 }
             }
         }
+
         val seconds = (currentTimeMillis() - time) / 1000.0
         val failureText = if (failureCount > 0) ". $failureCount tiles failed to download" else ""
         Log.i(TAG, "Downloaded $tileCount tiles (${downloadedSize / 1000}kB downloaded, ${cachedSize / 1000}kB already cached) in ${seconds.format(1)}s$failureText")
     }
 
-    private fun getDownloadTileSequence(bbox: BoundingBox): Sequence<Tile> =
+    private fun downloadTilesFlow(bbox: BoundingBox): Flow<DownloadResult> = channelFlow {
         /* tiles for the highest zoom (=likely current or near current zoom) first,
            because those are the tiles that are likely to be useful first */
-        (vectorTileProvider.maxZoom downTo 0).asSequence().flatMap { zoom ->
-            bbox.enclosingTilesRect(zoom).asTilePosSequence().map { Tile(zoom, it.x, it.y) }
+        (vectorTileProvider.maxZoom downTo 0).forEach { zoom ->
+            bbox.enclosingTilesRect(zoom).asTilePosSequence().forEach { (x, y) ->
+                launch {
+                    try {
+                        val result = downloadTile(zoom, x, y)
+                        val logText = if (result.alreadyCached) "in cache" else "downloaded"
+                        Log.v(TAG, "Tile $zoom/$x/$y $logText")
+                        send(result)
+                    } catch (e: Exception) {
+                        Log.w(TAG, "Error retrieving tile $zoom/$x/$y: ${e.message}")
+                        send(DownloadFailure)
+                    }
+                }
+            }
         }
+    }
 
-    private suspend fun downloadTile(zoom: Int, x: Int, y: Int): DownloadResult = suspendCancellableCoroutine { cont ->
+    private suspend fun downloadTile(zoom: Int, x: Int, y: Int): DownloadSuccess = suspendCancellableCoroutine { cont ->
         /* adding trailing "&" because Tangram-ES also puts this at the end and the URL needs to be
            identical in order for the cache to work */
         val url = vectorTileProvider.getTileUrl(zoom, x, y) + "&"
         val httpUrl = HttpUrl.parse(url)
         require(httpUrl != null) { "Invalid URL: $url" }
 
-        val builder = Request.Builder()
+        val request = Request.Builder()
             .url(httpUrl)
             .cacheControl(cacheConfig.cacheControl)
-        builder.header("User-Agent", ApplicationConstants.USER_AGENT + " / " + Version.userAgent())
-        val call = okHttpClient.newCall(builder.build())
+            .header("User-Agent", ApplicationConstants.USER_AGENT + " / " + Version.userAgent())
+            .build()
```

Comment:
This is mostly a matter of style, but I think it is clearer, since it does not requires double-checking what kind of builder it is.